_posts/2018-11-13-type-constraints-correctness-reuse.md

```diff
@@ -0,0 +1,65 @@
+
+Or maybe just take Int for performance reasons? Well, not all abstractions are free of cost, but the same reason we give for not using lower level languages applies - we want to develop software faster.
+
+Maybe we will use negative numbers in the future? Well, first the program must run in the present. Refactoring constraints is ok, the compiler will help.
+
+And will our program be correct if we do use Int? It might, but we will have to test the case when Int goes negative. This is what is meant by reducing implementation space. Using Nat achieves this, but Int will not. So we end up by having to write extra tests. We can see that in a dynamic language with no types, everything is unconstrained to some primitive types, thus all implementations must be tested.
+
+We've made a quick mental exercise on the cases above and we can see downsides and upsides as with most things. The bottom line is that if we are not solving a C10k like problem, we are probably better off building a more readable, smaller, well constrained program.
+
+Let's now take a look at `List` and `NonEmptyList` (from cats library).
+
+List represents something which may or may not have elements and can be constructed as such:
+- `List()` is the empty list
+- `List(x)` is the list of one element
+- `List(x, y, ...)` is the list of many elements
+
+NonEmptyList represents something which may never have the first case above.
+- `NonEmptyList(x)` is the list of one element
+- `NonEmptyList(x, y, ...)` is the list of many elements
+
+And it is actually a simple data type: `case class NonEmptyList[+A](head: A, tail: List[A])`.
+
+Also note that we can represent the empty list case with `Option[NonEmptyList[A]]` where None is the empty list.
+
+Now, we can choose to implement this simple fold using a List or a NonEmptyList.
+
+```
+def merge[A](x: A, y: A): A = ???
+
+def fromList[A](ls: List[A]): Option[A] =
+    ls.foldLeft(Option.empty[A])((b, a) => b.map(merge(_, a)))
+
+def fromNel[A](nel: NonEmptyList[A]): A =
+    nel.tail.foldLeft(nel.head)(merge)
+```
+Take a moment to appreciate that we removed the need for Option from the `fromNel` function implementation. We delayed the need to implement such behavior. This is due to List having an extra check compared to NonEmptyList: `isEmpty`. Our NonEmptyList function is more robust as it requires no testing for this case.
+
+Actually, isEmpty is equivalent to `NonEmptyList.fromList` which accepts a `List` and returns an `Option[NonEmptyList[A]]`. So anytime we are doing multiple isEmpty checks we should probably realize our program should be working with a NonEmptyList.
+
+Also note that `.filter` on a NonEmptyList will transform it back to a List, because it may remove all elements. So we should also consider consolidating filtering in our programs.
+
+As yet another example, let's say we are parsing some data into a case class and that data has a List field which is never empty.
```

Comment:
I think we can remove the "As yet" here also. The text flows nicely without it. Seems almost as if rushing using these time pronouns or wtv :)